graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWebSocketSubscriber.java

```diff
@@ -0,0 +1,121 @@
+import com.linecorp.armeria.common.websocket.WebSocketFrame;
+import com.linecorp.armeria.common.websocket.WebSocketWriter;
+
+@SuppressWarnings("ReactiveStreamsSubscriberImplementation")
+class GraphqlWebSocketSubscriber implements Subscriber<WebSocketFrame> {
+    private static final Logger logger = LoggerFactory.getLogger(GraphqlWebSocketSubscriber.class);
+    private final GraphqlWSSubProtocol graphqlWSSubProtocol;
+    private final WebSocketWriter outgoing;
+    @Nullable
+    Subscription subscription;
+
+    GraphqlWebSocketSubscriber(GraphqlWSSubProtocol graphqlWSSubProtocol, WebSocketWriter outgoing) {
+        this.graphqlWSSubProtocol = graphqlWSSubProtocol;
+        this.outgoing = outgoing;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (subscription != null) {
+            /*
+            A Subscriber MUST call Subscription.cancel() on the given Subscription after an
+            onSubscribe signal if it already has an active Subscription.
+             */
+            subscription.cancel();
+        }
+        subscription = s;
+        s.request(1);
+    }
+
+    /**
+     * Calling onSubscribe, onNext, onError or onComplete MUST return normally except when any provided
+     * parameter is null in which case it MUST throw a java.lang.NullPointerException to the caller, for all
+     * other situations the only legal way for a Subscriber to signal failure is by cancelling its
+     * Subscription. In the case that this rule is violated, any associated Subscription to the Subscriber
+     * MUST be considered as cancelled, and the caller MUST raise this error condition in a fashion that is
+     * adequate for the runtime environment.
+     */
+    @Override
+    public void onNext(WebSocketFrame webSocketFrame) {
+        logger.trace("onNext: {}", webSocketFrame);
+        if (subscription == null) {
+            return;
+        }
+        switch (webSocketFrame.type()) {
+            case TEXT:
+                // Parse the graphql-ws sub protocol. Maybe this could be done in a different thread so not
+                // to block the publisher?
+                graphqlWSSubProtocol.handle(webSocketFrame.text(), outgoing);
+                /*
+                It is RECOMMENDED that Subscribers request the upper limit of what they are able to process,
+                as requesting only one element at a time results in an inherently
+                inefficient "stop-and-wait" protocol.
+                 */
+                subscription.request(1);
+                break;
+            case PING:
+                outgoing.writePong();
+                subscription.request(1);
+                break;
+            case CLOSE:
+                subscription.cancel();
+                outgoing.close();
+                break;
+            // PONG is a noop
+            case PONG:
+                // These below will never happen?
+            case BINARY:
+            case CONTINUATION:
```

Comment:
Is there any reference on this in the graphql-ws spec?

If there isn't any specification:
- I feel like we can probably handle `BINARY` the same way that we handle `TEXT` by just using the `text()` method.
- For `CONTINUATION`, I guess it's not necessary to handle now and we can just log. I guess it would involve checking the `webSocketFrame.isFinalFragment()` flag and accumulating to a queue.